### DISCUSSION TITLE
cpg0037-oasis

### DISCUSSION THREAD

**Discussion by shntnu** [broadinstitute/cellpainting-gallery#92] [Dataset descriptions]
The data from Axiom will kick us off!

We will follow the process documented here https://github.com/broadinstitute/cellpainting-gallery/wiki/Data-Transfer-Workflow

Axiom team, could you please answer the following:

- [x] assay used (standard Cell Painting or describe the variation. If you would like to contribute data from a derivative assay it must be useable for morphological profiling in that it stains/labels multiple cellular compartments/organelles.)
- [x] approximate data size
- [x] components you wish to contribute (all are described in [data structure](https://broadinstitute.github.io/cellpainting-gallery/data_structure.html)) (major components: `images`, `analysis`, `backend`, `load_data_csv`, `profiles`. optional components: `pipelines`, `qc`, etc.). Note that `metadata` is required.
- [x] suggested top level project tag. Typically this is a 1 word summary of the project (e.g. cpg0011-lipocyteprofiler, cpg0016-jump, cpg0022-cmqtl) and sometimes also includes the last name of the first author (e.g. cpg0010-caie-drugresponse, cpg0028-kelley-resistance, cpg0031-caicedo-cmvip) ---> `cpg0037-oasis`

Then, please orient yourself to the prep in https://broadinstitute.github.io/cellpainting-gallery/contributing_to_cpg.html#preparing-for-data-deposition

---

For Broadies, here's our checklist from https://github.com/broadinstitute/cellpainting-gallery/wiki/Data-Transfer-Workflow

Champion: Shantanu SIngh
Dataset name: `cpg0037-oasis`

- [x] Contact a maintainer asking to deposit data, following the instructions in the [contribution docs](https://broadinstitute.github.io/cellpainting-gallery/contributing_to_cpg.html#contributing-data-to-the-gallery).
- [x] The Maintainer will assign a dataset identifier (e.g., `cpg0031`) by incrementing the identifier of the most recent dataset in the dataset [list](https://github.com/broadinstitute/cellpainting-gallery/blob/main/documentation/prefixes.md), and update the list with the new identifier.
- [x] The Maintainer will create a new [discussion](https://github.com/broadinstitute/cellpainting-gallery/discussions) using the dataset identifier as the title, which will be used to track the progress of this workflow
- [x] The Maintainer will identify an internal Champion for the project.
- [x] The Champion and the Contributor will discuss the structure of the data, ensuring it follows the guidelines discussed in the [contribution docs](https://broadinstitute.github.io/cellpainting-gallery/contributing_to_cpg.html#contributing-data-to-the-gallery).
- [x] The Champion and the Contributor will come up with an appropriate name for the dataset (e.g., `cpg0031-caicedo-cmvip`).
- [x] The Champion will request credentials for the Contributor by specifying the prefix (email [@leoank](https://github.com/leoank) and cc the Contributor). Each credential is for a single prefix on the staging bucket; multiple people can assume that same credential.
- [x] The Contributor will use provided credentials to upload data to the staging bucket, following [instructions](https://github.com/broadinstitute/cytoskel/blob/main/cytoskel/docs/access_cpg_staging.md) on accessing and using the credentials.
- [x] The Champion will verify data structure and give the go-ahead to transfer from the staging bucket to the production bucket.
- [x] A Maintainer will perform the transfer using an automated process with AWS S3 Batch.
- [ ] A Maintainer will disable the staging prefix, delete the data from staging, and delete credentials. Currently, tag [@leoank](https://github.com/leoank) for the last two steps; in the future, it could be any Maintainer.

**Comment by ktitterton** [comment]
Hi Shantanu, thanks for getting the ball rolling on this! Excited!

Confirming the assay used is the standard Cell Painting v3.0 dye set to stain primary human hepatocytes (PHHs). There are 68 x 384 well plates in the dataset. We use the whole plate and do not exclude any edge wells.  We image 15 sites (field of view) using the 40x objective per well. The field images are comprised of 5 fluorescent channels and 1 BF image. Each individual channel image is around 8 MB. Therefore it looks like the images will be around 20 TB, in sum. Images were acquired on an Operetta - we will also transfer the associated flat field correction (FFC) profiles and xml metadata for each plate.

In addition to the images, we are also happy to share our in-house analysis QC pipeline results, including:

1. Nuclei segmentation masks (generated using a method which has been optimized for the PHH cells) and associated object property features (area, eccentricity, etc)
2. Basic whole-image statistical features
3. Extracted image embeddings for each channel image

Leveraging the embeddings has proven particularly insightful. We have seen that ~70% of the compounds in this dataset produce image embeddings that are significantly different  than the DMSO vehicle control embeddings in the same plate, and could be considered bioactive.

**Comment by shntnu** [comment]
Thanks for all these details @ktitterton!

Let's start with image data first. Do you anticipate any issues in structuring the data like this? 
https://broadinstitute.github.io/cellpainting-gallery/data_structure.html#images-folder-structure

As for the remaining components, I suspect they will all go into one of the existing folders in `workspace` or `workspace_dl`. For each of these components, 

- Nuclei segmentation masks (generated using a method which has been optimized for the PHH cells) and associated object property features (area, eccentricity, etc)
- Basic whole-image statistical features
- Extracted image embeddings for each channel image

1. is it possible to arrange them in the `<batch>/<plate>/` structure?
1. Assuming yes, can you elaborate on how they are currently organized under `<plate>`?

And then finally, would it be possible for you to create metadata for each of these plates, as detailed in https://broadinstitute.github.io/cellpainting-gallery/data_structure.html#arrayed-metadata?

**Comment by ktitterton** [comment]
Hi Shantanu,

Currently our data are indeed under a `<batch>/<plate>/ ` folder structure. For each batch, we have separate folders for the image data that comes off the microscope, and the transformed and associated metadata.

The raw image data follows the standard PerkinElmer/Revvity format:

- prod_25
  - 41002687__2024-05-20T00_39_21-Measurement 1
    - FCC_Profile
    - Images
      - Index.xml
      - r01c01f01p01-ch1sk1fk1fl1.tiff


The associated metadata and features folder structure is:

- prod_25
  - plate_41002687
    - biochem.parquet
    - metadata.parquet
    - segmentation_masks
    - qc
      - prod
        - dinov2_b_fieldnorm
        - image_metrics
        - segmentation

 
Some misc. notes:
- `segmentation_masks` contains a folder for each well (capital and not zero padded, A1-P24). 
  - Within each well folder, there are zero-indexed field folders for fields 0-14.
  - Within each field folder lives the `.npy` mask file for that particular image.
- The subfolders under `prod` containing embeddings (`dinov2_fieldnorm`), `image_metrics`, and `segmentation` image features all contain parquet files.
- `biochem.parquet` contains a `well_id` column as a unique identifier, as this data is collected on a well level.
  - All of the other data use an `image_id` unique identifier. 
  - `image_id` follows the format `assayworks_prod_25_p=plate_41002687_r=0_c=0_f=0,` where `p` is plate, `r` is row, `c` is column, and `f` is field.
  - `well_id` is the same, except it drops the final field (_f=0) substring.
- `metadata.parquet` contains information regarding the [compound, dose] perturbations for each `image_id`, as well as batch and plate.


Yes, we are happy to create metadata following the specifications. Is this to be provided in addition to the data as it currently exists, described above? 

tagging my colleague Alex @cabreraalex who is our resident data wizard and much more expert in these matters than myself!

**Reply by shntnu** [reply]
Thanks for all these details

> Is this to be provided in addition to the data as it currently exists, described above?

Yes please

(Please see my next comment)

**Comment by ktitterton** [comment]
The Axiom team just chatted a bit internally, and we were wondering if it would be most helpful to go ahead and copy over the data as it currently exists into an s3 bucket, and share it with y'all - that way you can let us know exactly what should be added, deleted, or rearranged. Let us know if this would be a good next step! Excited!

**Reply by shntnu** [reply]
Great idea! I will ping @leoank via email to send you credentials

**Comment by shntnu** [comment]
@ktitterton The images are structured exactly per specs, so all set there. The metadata.parquet looks good but will eventually need to be wrangled to follow the structure, but that can wait.

Next, can you please provide the following for 1-2 plates?

```
- prod_25
  - plate_41002687
    - biochem.parquet
    - metadata.parquet
    - segmentation_masks
    - qc
      - prod
        - dinov2_b_fieldnorm
        - image_metrics
        - segmentation
```

Please upload that to `s3://staging-cellpainting-gallery/cpg0037-oasis/axiom/workspace/scratch/` for now

**Reply by jessica-ewald** [reply]
Hi @shntnu and @ktitterton - are the images ready for @ErinWeisbart to start processing with CellProfiler, or are we waiting on some more checks/data?

**Reply by shntnu** [reply]
The images look good to me but @ktitterton can you confirm these are ALL the images?

**Reply by cabreraalex** [reply]
Hi ya'll! Confirmed all the images are in, and I transferred the rest of the metadata into the scratch workspace folder. LMK if it all looks alright or if anything looks off.

**Comment by jessica-ewald** [comment]
Hi @cabreraalex - thanks for transferring everything, we've kicked off the image processing on our end. 

Do you, @ktitterton, or @shntnu know the plan for transferring the Dino features? In other projects, we've created additional folders to hold alternative features, for example in this case we have 'images' for the raw .tiff, 'workspace' for CellProfiler features, and 'workspace_dl' for some deep learning-derived features:

<img width="577" alt="image" src="https://github.com/user-attachments/assets/b67a82ab-39d2-49b9-bff7-4e70d4bf62f8">

I imagine that we'll want to do something like this as well. Has it already been discussed?

**Reply by jessica-ewald** [reply]
(I see the comment about with @shntnu's mock file structure, but don't see that reflected in the CellPainting gallery)

**Reply by cabreraalex** [reply]
Hi Jessica! The embeddings should be there in a path like
`cpg0037-oasis/axiom/workspace/scratch/prod_25/plate_41002692/dinov2_b_fieldnorm.parquet`

**Reply by jessica-ewald** [reply]
Oh - it's possible that I'm getting ahead of things :) I thought that the data was finished syncing but Ank just told me that it'll take ~10 more hours. That's probably why I can't see it.

**Reply by cabreraalex** [reply]
Ah interesting it should already be in s3 but maybe syncing locally?

**Reply by jessica-ewald** [reply]
We have a staging bucket where Ank/Shantanu/Erin would have reviewed, and then it gets synced to the "real" cpg that the rest of the lab can access. So I think the sync that Ank is referring to is between those two locations.

**Comment by ErinWeisbart** [comment]
@ktitterton 
I'm getting started on processing the images however, I see in your comment above that the PE microscope exported `.xml` metadata files with image acquisition and I am not seeing them in the Cell Painting Gallery. The `.xml` files are important for our first step of generating load data information. 
It looks like the image folders were not uploaded in the same structure as they are exported from the microscope (as you posted above):
```
- prod_25
  - 41002687__2024-05-20T00_39_21-Measurement 1
    - FCC_Profile
    - Images
       - Index.xml
       - r01c01f01p01-ch1sk1fk1fl1.tiff
```
But instead were uploaded with image files directly in the plate folder:
```
- prod_25
  - plate_41002892
       - r01c01f01p01-ch1sk1fk1fl1.tiff
```

How might we go about getting the `Index.xml` files?
Are you now able to upload them to the plate folder? i.e.
```
- prod_25
  - plate_41002892
      - Index.xml
      - r01c01f01p01-ch1sk1fk1fl1.tiff
```

**Reply by cabreraalex** [reply]
Hi Erin! I can upload the xml, will update when done

**Reply by cabreraalex** [reply]
Should be there now - prod 30 only has half the images (only half were OASIS), lmk if that causes an issue

**Comment by ErinWeisbart** [comment]
@ktitterton @cabreraalex 
I'm starting to generate load_data.csv's for the images. I've found that some of the Index.xml files are Harmony V6 and some are Harmony V7. I want to confirm that this is expected? This generally indicates that acquisition was done on separate microscopes - are there any known differences between the microscopes?

I also want to confirm that channel mapping I've parsed below is correct and is consistent across all plates and all batches?

**channels:**
    HOECHST 33342: OrigDNA
    WGA555/Phalloidin568: OrigAGP
    Mito_641: OrigMito
    PhenoVue Fluor 488: OrigER
    CP_Nuclei acid_512: OrigRNA
    Brightfield: OrigBrightfield

**channelid:**
    1: OrigDNA
    2: OrigER
    3: OrigAGP
    4: OrigRNA
    5: OrigMito
    6: OrigBrightfield

**Reply by ErinWeisbart** [reply]
Upon closer inspection, there are only 2 .xml files that were generated with Harmony V7, though my questions still stand.
`prod_25/plate_41002700/Index.xml`
`prod_25/plate_41002698/Index.xml`

**Reply by ErinWeisbart** [reply]
Queried version with:
```
import xml.etree.ElementTree as ET
import os

countv6 = 0
countv7 = 0
prefix = '/Users/eweisbar/Desktop/axiom/images/'
for batch in ['25','26','27','30']:
    plate_folders = os.listdir(os.path.join(prefix,f'prod_{batch}'))
    for folder in plate_folders:
        if os.path.exists(os.path.join(prefix,f'prod_{batch}',folder)):
            index_file = os.path.join(prefix,f'prod_{batch}',folder,'Index.xml')
            namespaces = dict([node for _, node in ET.iterparse(index_file, events=['start-ns'])])
            if "HarmonyV6" in namespaces['']:
                countv6 +=1
            elif "HarmonyV7" in namespaces['']:
                countv7 +=1
                print (index_file)
            else:
                print('error', index_file)
print (countv6,countv7)
```

**Comment by ktitterton** [comment]
Hi Erin, 

Yes, I can confirm there are two different Harmony versions and two different microscopes in the dataset. The vast majority of the data was acquired on O1, with a very small minority on O2. There was an unfortunate automation error in prod_25, so there are some technical artifacts in this batch, including the microscope switch for a couple of plates. In our unbiased image clustering, we have noticed some patterning by microscope in high-tox conditions. In general, the O2 microscope seems to generate images which have slightly more texture -  but the raw image histograms are otherwise extremely similar.

Yup, I can also confirm that channel mapping looks correct! Please do let us know if you have any other questions- no question is too small. We are all very much looking forward to seeing your results sometime soon! Thanks for checking in

**Comment by ErinWeisbart** [comment]
@ktitterton 
I've noticed discrepancies in image file number and I want to confirm that is expected. The majority of image folders contain 34561 files (15 sites (fov) * 384 wells * 6 channels + Index.xml). Major exceptions are:
`prod_25`: plate_41002687 (14401 files)
`prod_26`: plate_41002891 (14401 files)
`prod_27`: plate_41002908 (14401 files)
`prod_30`: none are complete, sizes range from 14311 to 22951 files

There are several that are off by 6 files (34555 files) which would be a single field of view and I am not concerned about these, but I will note them here for the sake of completeness.
`prod_25`: plate_41002688 , plate_41002689, plate_41002693
`prod_26`: plate_41002889
`prod_27`: plate_41002899

**Reply by cabreraalex** [reply]
Hi Erin! for prod 25 - 27 those missing plates are the reference plates, didn't upload the images.
For prod 30, only half the batch was OASIS compounds, hence the missing images - they should also be missing from the biochem and metadata parquets.

**Reply by ErinWeisbart** [reply]
Great, thanks for confirming @cabreraalex !

**Comment by jessica-ewald** [comment]
Hi @cabreraalex and @ktitterton - I'm not seeing any positive controls or negative controls in the Dino embeddings (and wondering if @ErinWeisbart has any in the images / resulting CellProfiler features). 

After compiling the metadata.parquet across all plates in locations like [this](https://cellpainting-gallery.s3.amazonaws.com/index.html#cpg0037-oasis/axiom/workspace/scratch/prod_25/plate_41002693/), I don't see any compounds with the name "DMSO" or "dimethyl sufoxide", and no compounds with a concentration of 0. 

Can you let me know if you either A) didn't collect poscons/negcons on each plate or B) just didn't include them in the data that you uploaded? Poscons I can live without, but the DMSO samples are pretty crucial for the normalization and the concentration-response analysis.

**Reply by cabreraalex** [reply]
Hi Jessica! I likely missed uploading the DMSO controls let me see if I can include them.

**Reply by cabreraalex** [reply]
Could you link me to the bucket you're looking at?

**Reply by cabreraalex** [reply]
(just kidding I see it above)

**Reply by cabreraalex** [reply]
Alright looks like the images for DMSO are uploaded but I excluded it from all the metadata. 

Could I reupload to the same bucket and it should work?

**Reply by jessica-ewald** [reply]
I'll tag @leoank here - I believe that you originally uploaded to a staging bucket which was synced with the real CPG after review. I assume uploading to the same place that you originally uploaded makes sense, but maybe Ank can confirm? I don't believe you'll be able to upload directly to the bucket that I linked.

**Reply by cabreraalex** [reply]
Cool I just uploaded updated metadata and biochem to the staging bucket and am uploading the embeddings and other features now

**Reply by jessica-ewald** [reply]
Great, thanks so much!

**Reply by shntnu** [reply]
@jessica-ewald I'm copying over updated metadata.parquet and biochem.parquet files. You'd need to reverify that everything looks good.

<details>

```sh
#!/bin/bash

# Function to check if a file exists in S3
file_exists_in_s3() {
    if aws s3 ls "$1" >/dev/null 2>&1; then
        return 0
    else
        return 1
    fi
}

SOURCE_DIR="s3://staging-cellpainting-gallery/cpg0037-oasis/axiom/workspace/scratch/"
SOURCE_BUCKET="s3://staging-cellpainting-gallery/"
DEST_BUCKET="s3://cellpainting-gallery/"

# List all biochem.parquet files in the source bucket
aws s3 ls --recursive "$SOURCE_DIR" | grep -E '(biochem|metadata)\.parquet' | sort | while read -r line; do
    # Extract the file path
    file_path=$(echo "$line" | awk '{print $4}')

    # Construct the source and destination paths
    source_path="$SOURCE_BUCKET$file_path"
    dest_path="$DEST_BUCKET$file_path"

    # Check if the source file exists
    if file_exists_in_s3 "$source_path"; then
        # Copy the file
        aws s3 cp "$source_path" "$dest_path"
    else
        echo "Source file does not exist: $source_path"
    fi

done

```

**Comment by shntnu** [comment]
Hi folks, our [plan](https://github.com/broadinstitute/cellpainting-gallery/discussions/92#discussioncomment-10204518) was to upload Axiom's processed data as-is and then re-organize later.

Let's do that now

We will follow the instructions here
https://broadinstitute.github.io/cellpainting-gallery/data_structure.html#workspace-dl-folder-structure

This is the current structure, under `workspace/scratch`


```
- prod_25
  - plate_41002687
    - biochem.parquet
    - dinov2_b_fieldnorm.parquet
    - image_metrics.parquet
    - metadata.parquet
    - segmentation.parquet
```

Proposed structure

Under `workspace`:

```
- segmentation
  - prod_25
    - plate_41002687
      - segmentation.parquet
- qc
  - prod_25
    - plate_41002687
      - image_metrics.parquet
- metadata
  - prod_25
    - plate_41002687
      - metadata.parquet
      - biochem.parquet
```

The `metadata` structure doesn't follow our standard structure, but let's go with this for now, and fix that in the next iteration

Then under `workspace_dl`:

```
- profiles
  - dinov2_b_fieldnorm_xxx
    - prod_25
      - plate_41002687
        - plate_41002687.parquet
```

- `dinov2_b_fieldnorm_xxx` --> for `xxx`, we need some unique identified of the model, like the hash of the model. For example, for EffNetv2 embeddings, we've named it `efficientnet_v2_imagenet1k_s_feature_vector_2_ec756ff`. Let us know what you'd like to tag it with @cabreraalex @ktitterton 
- `dinov2_b_fieldnorm.parquet` will be renamed `plate_41002687.parquet`

Additionally, @jessica-ewald has new CNN-based from Ray Jones and Bram Gorissen; these would be structured like this:

```
- profiles
  - cellpainting_cnn_retrained_xxx
    - prod_25
      - plate_41002687
        - plate_41002687.parquet
```

Does all this sound ok, @ktitterton @cabreraalex @jessica-ewald? 

We (Broad) should be able to help complete this task but might ping some of you for help

**Reply by jessica-ewald** [reply]
Sounds good mostly, the only edit I have is that most "metadata.parquet" files (I believe all but one) are now empty parquets and biochem.parquet actually contains all of the relevant details. 

I believe this change was made when @cabreraalex helped out with compliance with OASIS guidelines

**Reply by shntnu** [reply]
Ah yes, I see this as the only non-empty metadata file @cabreraalex 

```
2024-09-18 09:38:50     580430 cpg0037-oasis/axiom/workspace/scratch/prod_25/plate_41002688/metadata.parquet
```

**Reply by cabreraalex** [reply]
let me double check the metadata.parquets!

**Reply by cabreraalex** [reply]
Oops yup bug on my end, can you check them now?

**Reply by cabreraalex** [reply]
re: reorganization plan that sounds great to me. I'll let ya'll lead since the feedback loop will be quicker, but let me know if I can help out with anything/if anything doesn't make sense.

**Reply by shntnu** [reply]
No worries; syncing over now

@cabreraalex can you address this q:

> dinov2_b_fieldnorm_xxx --> for xxx, we need some unique identified of the model, like the hash of the model. For example, for EffNetv2 embeddings, we've named it efficientnet_v2_imagenet1k_s_feature_vector_2_ec756ff. Let us know what you'd like to tag it with

**Reply by cabreraalex** [reply]
maybe '825c11' if I understand correctly? That's a hash for https://github.com/facebookresearch/dinov2/blob/main/dinov2/configs/train/vitl14.yaml which is the train config for the dino version we use.

**Reply by shntnu** [reply]
Thanks for the hash

What's more appropriate: `dinov2_vitl14_825c11`  or `dinov2_b_fieldnorm_825c11`?

**Reply by cabreraalex** [reply]
let's keep fieldnorm since we do normalize by field and should make it explicit

**Reply by cabreraalex** [reply]
we can add both if we want too: dinov2_b_vitl14_fieldnorm_825c11 go crazy

**Reply by shntnu** [reply]
Great, sync is done!

---


Sync code:

<details>

```py
import boto3
import logging
import argparse
from botocore.exceptions import ClientError

# Set up logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
)

# Initialize S3 client
s3 = boto3.client("s3")

# Define S3 bucket and paths
BUCKET_NAME = "cellpainting-gallery"
SOURCE_PREFIX = "cpg0037-oasis/axiom/workspace/scratch/"
WORKSPACE_PREFIX = "cpg0037-oasis/axiom/workspace/"
WORKSPACE_DL_PREFIX = "cpg0037-oasis/axiom/workspace_dl/"

# List of production batches
PROD_BATCHES = ["prod_25", "prod_26", "prod_27", "prod_30"]

# Hardcoded model identifier
MODEL_IDENTIFIER = "dinov2_b_vitl14_fieldnorm_825c11"


def s3_copy_object(src_key, dst_key, dry_run=False):
    """Copy object within the same S3 bucket."""
    if dry_run:
        logging.info(f"[DRY RUN] Would copy {src_key} to {dst_key}")
    else:
        try:
            s3.copy_object(
                Bucket=BUCKET_NAME,
                CopySource={"Bucket": BUCKET_NAME, "Key": src_key},
                Key=dst_key,
            )
            logging.info(f"Copied {src_key} to {dst_key}")
        except ClientError as e:
            logging.error(f"Error copying {src_key} to {dst_key}: {e}")


def reorganize_workspace(dry_run=False):
    """Reorganize files in the workspace directory for all production batches."""
    for prod_batch in PROD_BATCHES:
        batch_prefix = f"{SOURCE_PREFIX}{prod_batch}/"
        logging.info(f"Processing batch: {prod_batch}")

        paginator = s3.get_paginator("list_objects_v2")
        for page in paginator.paginate(Bucket=BUCKET_NAME, Prefix=batch_prefix):
            for obj in page.get("Contents", []):
                src_key = obj["Key"]
                filename = src_key.split("/")[-1]
                plate = src_key.split("/")[-2]

                if filename == "segmentation.parquet":
                    dst_key = f"{WORKSPACE_PREFIX}segmentation/{prod_batch}/{plate}/{filename}"
                elif filename == "image_metrics.parquet":
                    dst_key = f"{WORKSPACE_PREFIX}qc/{prod_batch}/{plate}/{filename}"
                elif filename == "metadata.parquet" or filename == "biochem.parquet":
                    dst_key = (
                        f"{WORKSPACE_PREFIX}metadata/{prod_batch}/{plate}/{filename}"
                    )
                else:
                    continue

                s3_copy_object(src_key, dst_key, dry_run)


def reorganize_workspace_dl(dry_run=False):
    """Reorganize files in the workspace_dl directory for all production batches."""
    for prod_batch in PROD_BATCHES:
        batch_prefix = f"{SOURCE_PREFIX}{prod_batch}/"
        logging.info(f"Processing batch: {prod_batch}")

        paginator = s3.get_paginator("list_objects_v2")
        for page in paginator.paginate(Bucket=BUCKET_NAME, Prefix=batch_prefix):
            for obj in page.get("Contents", []):
                src_key = obj["Key"]
                filename = src_key.split("/")[-1]
                plate = src_key.split("/")[-2]

                if filename == "dinov2_b_fieldnorm.parquet":
                    dst_key = f"{WORKSPACE_DL_PREFIX}profiles/{MODEL_IDENTIFIER}/{prod_batch}/{plate}/{plate}.parquet"
                    s3_copy_object(src_key, dst_key, dry_run)


def main():
    parser = argparse.ArgumentParser(description="Reorganize S3 files")
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Perform a dry run without making actual changes",
    )
    args = parser.parse_args()

    if args.dry_run:
        logging.info("Performing a dry run. No actual changes will be made.")
    else:
        logging.info("Starting S3 file reorganization")

    reorganize_workspace(args.dry_run)
    reorganize_workspace_dl(args.dry_run)

    if args.dry_run:
        logging.info("Dry run completed. No actual changes were made.")
    else:
        logging.info("S3 file reorganization completed")


if __name__ == "__main__":
    main()

```
</details>

Validation code:

<details>

```py
import boto3
import logging
from collections import defaultdict

# Set up logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
)

# Initialize S3 client
s3 = boto3.client("s3")

# Define S3 bucket and paths
BUCKET_NAME = "cellpainting-gallery"
WORKSPACE_PREFIX = "cpg0037-oasis/axiom/workspace/"
WORKSPACE_DL_PREFIX = "cpg0037-oasis/axiom/workspace_dl/"

# List of production batches
PROD_BATCHES = ["prod_25", "prod_26", "prod_27", "prod_30"]

# Hardcoded model identifier
MODEL_IDENTIFIER = "dinov2_b_vitl14_fieldnorm_825c11"


def list_s3_objects(prefix):
    """List all objects in S3 with the given prefix."""
    objects = []
    paginator = s3.get_paginator("list_objects_v2")
    for page in paginator.paginate(Bucket=BUCKET_NAME, Prefix=prefix):
        if "Contents" in page:
            objects.extend(page["Contents"])
    return objects


def verify_workspace_structure():
    """Verify the structure of files in the workspace directory."""
    expected_structure = {
        "segmentation": ["segmentation.parquet"],
        "qc": ["image_metrics.parquet"],
        "metadata": [
            "metadata.parquet",
            "biochem.parquet",
        ],
    }

    errors = defaultdict(list)

    for prod_batch in PROD_BATCHES:
        for folder, expected_files in expected_structure.items():
            prefix = f"{WORKSPACE_PREFIX}{folder}/{prod_batch}/"
            objects = list_s3_objects(prefix)

            plates = set()
            for obj in objects:
                key = obj["Key"]
                parts = key.split("/")
                if (
                    len(parts) != 7
                ):  # cpg0037-oasis/axiom/workspace/folder/prod_batch/plate/file
                    errors[prod_batch].append(f"Unexpected key structure: {key}")
                    continue

                plate = parts[-2]
                filename = parts[-1]
                plates.add(plate)

                if filename not in expected_files:
                    errors[prod_batch].append(f"Unexpected file in {folder}: {key}")

            # Check if all plates have all expected files
            for plate in plates:
                for expected_file in expected_files:
                    expected_key = f"{prefix}{plate}/{expected_file}"
                    if not any(obj["Key"] == expected_key for obj in objects):
                        errors[prod_batch].append(f"Missing file: {expected_key}")

    return errors


def verify_workspace_dl_structure():
    """Verify the structure of files in the workspace_dl directory."""
    errors = defaultdict(list)

    for prod_batch in PROD_BATCHES:
        prefix = f"{WORKSPACE_DL_PREFIX}profiles/{MODEL_IDENTIFIER}/{prod_batch}/"
        objects = list_s3_objects(prefix)

        plates = set()
        for obj in objects:
            key = obj["Key"]
            parts = key.split("/")
            if (
                len(parts) != 8
            ):  # cpg0037-oasis/axiom/workspace_dl/profiles/dinov2_b.../prod_batch/plate/file
                errors[prod_batch].append(
                    f"Unexpected key structure: {key}. len(parts) = {len(parts)}"
                )
                continue

            plate = parts[-2]
            filename = parts[-1]
            plates.add(plate)

            if filename != f"{plate}.parquet":
                errors[prod_batch].append(f"Unexpected filename: {key}")

        # Check if all plates have their corresponding file
        for plate in plates:
            expected_key = f"{prefix}{plate}/{plate}.parquet"
            if not any(obj["Key"] == expected_key for obj in objects):
                errors[prod_batch].append(f"Missing file: {expected_key}")

    return errors


def main():
    logging.info("Starting S3 file structure verification")

    workspace_errors = verify_workspace_structure()
    workspace_dl_errors = verify_workspace_dl_structure()

    if not workspace_errors and not workspace_dl_errors:
        logging.info("Verification completed. All files are correctly structured.")
    else:
        logging.error("Verification completed. Errors were found:")
        for prod_batch, errors in workspace_errors.items():
            for error in errors:
                logging.error(f"{prod_batch}: {error}")
        for prod_batch, errors in workspace_dl_errors.items():
            for error in errors:
                logging.error(f"{prod_batch}: {error}")


if __name__ == "__main__":
    main()

```
</details>

Log:

[log.txt](https://github.com/user-attachments/files/17101732/log.txt)

**Comment by shntnu** [comment]
For our notes -- we also have an internal email thread discussing this "Updating metadata.parquet files / blocklists"
We are all set with the data upload, so we will remove it from our staging bucket. 
Thanks all!
